File: scripts/content-engine/youtube_ads.js
#!/usr/bin/env node
/* youtube_ads.js — read and set AD BREAKS and AD FORMATS for every video on the channel.
 *
 * Kevin, 20 Sep 2026: "every single video has the ads switched on so that they can generate money".
 * The master monetisation switch (youtube_studio.py) was already On everywhere. What was off was the
 * ad in the MIDDLE: 817 of 888 videos over 8 minutes had mid-roll ads unticked, and YouTube had already
 * worked out where the break goes on 695 of them.
 *
 * YouTube's Data API has no ad-break setting, so this drives Studio's own endpoint:
 *   read : POST /youtubei/v1/creator/list_creator_videos  with adSettings in the mask
 *   write: POST /youtubei/v1/video_manager/metadata_update
 *
 * The write carries a BotGuard `attestationResponseData` that the page mints. WITHOUT IT THE WRITE
 * RETURNS 200 AND CHANGES NOTHING (proved 20 Sep 2026 on Rs8xHbD5miQ: 200 OK, mid-roll still false).
 * So the first video of every batch is saved through the real UI, which mints an attestation, and the
 * rest replay that same attestation with only encryptedVideoId swapped. Every write is read back.
 *
 *   youtube_ads.js audit [--out FILE]              every video, its ad breaks and ad formats
 *   youtube_ads.js fix --ids FILE [--dry]          mid-roll (+pre/post-roll) On for those video ids
 *   youtube_ads.js legacy --ids FILE [--dry]       convert pre-2024 videos off legacy ad settings
 *   youtube_ads.js verify --ids FILE               read back only
 */
const path = require('path'), os = require('os'), fs = require('fs');
const { chromium } = require('/Users/kevinbrittain/Projects/leadership-dashboard/node_modules/playwright');

const UA = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/140.0.0.0 Safari/537.36';
const PROFILE = 'default';
const CHANNEL = 'UCKvcFYC2y4FAeTDS20_JXFw';
const LIST_URL = (tab) => `https://studio.youtube.com/channel/${CHANNEL}/videos/${tab}`;
const EARN_URL = (id) => `https://studio.youtube.com/video/${id}/monetization/ads`;
const MIDROLL_LABEL = 'Show mid-roll ads during my video';
// Mid-roll needs 8 minutes of video; YouTube hides the checkbox below that (read in Studio 20 Sep 2026).
const MIDROLL_MIN_SECONDS = 480;
const REMINT_AFTER = 120;        // a fresh attestation well inside its 6-hour TTL
// Two hours. A full-channel backfill is ~900 videos at two writes each plus a paged read-back per
// round; one hour was not enough headroom for a retry round (20 Sep 2026).
const WATCHDOG_MS = 7200000;

function args() {
  const a = {}; const v = process.argv.slice(2); a.mode = v[0];
  for (let i = 1; i < v.length; i++) {
    if (v[i] === '--ids') a.ids = v[++i];
    else if (v[i] === '--out') a.out = v[++i];
    else if (v[i] === '--dry') a.dry = true;
    else if (v[i] === '--limit') a.limit = parseInt(v[++i], 10);
  }
  return a;
}

async function open(headless = true) {
  const dir = path.join(os.homedir(), '.config', 'od', 'agent-browser', PROFILE);
  const ctx = await chromium.launchPersistentContext(dir, {
    headless, viewport: { width: 1400, height: 1000 }, channel: 'chrome',
    ignoreDefaultArgs: ['--enable-automation'], userAgent: UA });
  const page = ctx.pages()[0] || await ctx.newPage();
  page.on('dialog', (d) => d.accept().catch(() => {}));
  return { ctx, page };
}

async function settle(page) {
  await page.waitForTimeout(14000);
  const skip = page.getByText('SKIP TO YOUTUBE STUDIO', { exact: true }).first();
  if (await skip.count()) { await skip.click().catch(() => {}); await page.waitForTimeout(9000); }
}

/* ---------- read ---------- */

// Every video in one list tab. `adSettings` is added to the page's own mask, so the shape always
// matches what Studio itself reads. Paged to the end: a partial read would under-report the backlog.
async function listTab(page, tab) {
  let cap = null;
  const grab = (r) => { if (!cap && /list_creator_videos/.test(r.url())) cap = { url: r.url(), headers: r.headers(), body: r.postData() }; };
  page.on('request', grab);
  await page.goto(LIST_URL(tab), { waitUntil: 'domcontentloaded' });
  await settle(page);
  page.off('request', grab);
  if (!cap) throw new Error('Studio never issued list_creator_videos for ' + tab);
  return page.evaluate(async (cap) => {
    const base = JSON.parse(cap.body);
    base.mask.adSettings = { all: true };
    base.mask.monetizedStatus = true;
    const hdr = {}; for (const [k, v] of Object.entries(cap.headers)) if (!/^:|^host$|^content-length$/i.test(k)) hdr[k] = v;
    const rows = []; let token = null;
    for (let page_ = 0; page_ < 300; page_++) {
      const b = Object.assign({}, base, { pageSize: 100 });
      if (token) b.pageToken = token; else delete b.pageToken;
      const r = await fetch(cap.url, { method: 'POST', headers: hdr, body: JSON.stringify(b), credentials: 'include' });
      if (!r.ok) throw new Error('list_creator_videos http ' + r.status);
      const j = await r.json();
      const vids = j.videos || [];
      for (const v of vids) rows.push({ id: v.videoId, title: (v.title || '').slice(0, 90),
        published: v.timePublishedSeconds, seconds: Math.round((v.videoDurationMs || 0) / 1000),
        privacy: v.privacy, monetization: v.monetization, adSettings: v.adSettings });
      token = j.nextPageToken;
      if (!token || !vids.length) break;
    }
    return rows;
  }, cap);
}

async function audit(page) {
  const long = await listTab(page, 'upload');
  const shorts = await listTab(page, 'short');
  return { long, shorts };
}

/* ---------- write ---------- */

// One genuine UI save. It is what mints the attestation the replays reuse, and it fixes a real video.
async function mint(page, videoId) {
  let upd = null;
  const grab = (r) => { if (/metadata_update/.test(r.url())) upd = { url: r.url(), headers: r.headers(), body: r.postData() }; };
  page.on('request', grab);
  try {
    await page.goto(EARN_URL(videoId), { waitUntil: 'domcontentloaded' });
    await page.waitForTimeout(14000);
    const lit = page.locator('ytcp-checkbox-lit').filter({ hasText: MIDROLL_LABEL }).first();
    // A video under 8 minutes carries no mid-roll checkbox at all. That is YouTube's rule, not a fault:
    // it is reported as not-eligible so a short episode never sits in the backlog for ever (20 Sep 2026).
    if (!(await lit.count())) { const e = new Error('not eligible for mid-roll: ' + videoId); e.notEligible = true; throw e; }
    const box = lit.locator('#checkbox').first();
    // Already ticked: there is nothing to save, so Save stays disabled. That is a video that is DONE,
    // not a video that failed — treating it as a failure would leave it "failed" on every future run.
    if ((await box.getAttribute('aria-checked')) === 'true') { const e = new Error('already on: ' + videoId); e.alreadyOn = true; throw e; }
    await lit.scrollIntoViewIfNeeded();
    const bb = await box.boundingBox();
    if (!bb) throw new Error('mid-roll checkbox not on screen for ' + videoId);
    await page.mouse.click(bb.x + bb.width / 2, bb.y + bb.height / 2);
    await page.waitForTimeout(1800);
    if ((await box.getAttribute('aria-checked')) !== 'true') throw new Error('mid-roll did not tick on ' + videoId);
    const save = page.locator('ytcp-button:has-text("Save"):not([disabled]), #save-button:not([disabled])').first();
    if (!(await save.count())) throw new Error('Save never enabled on ' + videoId);
    await save.click({ timeout: 25000 });
    await page.waitForTimeout(9000);
  } finally { page.off('request', grab); }
  if (!upd) throw new Error('no metadata_update captured from ' + videoId);
  return upd;
}

// What a write asks for. MIDROLL is the modern case; LEGACY_FORMATS is the first half of converting a
// pre-2024 video off "legacy ad settings", which is the change YouTube warns cannot be reversed
// (Kevin approved it, 20 Sep 2026). Field names read off Studio's own two saves on xxYZcW8SFQk.
const MIDROLL = { adBreaks: { newHasPrerolls: 'ENABLED', newHasMidrollAds: 'ENABLED',
                              newHasPostrolls: 'ENABLED', newAutoMidrollEnabled: 'ENABLED' } };
const LEGACY_FORMATS = { adFormats: { newHasSkippableVideoAds: 'ENABLED', newHasNonSkippableVideoAds: 'ENABLED',
                                      newDisableDisplayAds: 'DISABLED' },
                         adBreaks: { newHasPrerolls: 'ENABLED', newHasPostrolls: 'ENABLED' } };

// Replay the minted write for one video. A 200 here is NOT proof: the caller reads every id back.
async function replay(page, upd, videoId, adSettings) {
  return page.evaluate(async ({ upd, videoId, adSettings }) => {
    const hdr = {}; for (const [k, v] of Object.entries(upd.headers)) if (!/^:|^host$|^content-length$/i.test(k)) hdr[k] = v;
    const b = JSON.parse(upd.body);
    b.encryptedVideoId = videoId;
    b.adSettings = adSettings;
    const r = await fetch(upd.url, { method: 'POST', headers: hdr, body: JSON.stringify(b), credentials: 'include' });
    return { status: r.status };
  }, { upd, videoId, adSettings });
}

// The legacy conversion, through the UI, on one video. Two saves: the banner's irreversible "Update",
// which turns video ads on, and then mid-roll if the video is long enough. Either save mints an
// attestation the replays can reuse.
async function mintLegacy(page, videoId) {
  let upd = null;
  const grab = (r) => { if (/metadata_update/.test(r.url())) upd = { url: r.url(), headers: r.headers(), body: r.postData() }; };
  page.on('request', grab);
  try {
    await page.goto(EARN_URL(videoId), { waitUntil: 'domcontentloaded' });
    await page.waitForTimeout(14000);
    const banner = page.locator('ytcp-button:has-text("Update")').first();
    if (!(await banner.count())) { const e = new Error('not a legacy video: ' + videoId); e.notLegacy = true; throw e; }
    await banner.click({ timeout: 20000 });
    await page.waitForTimeout(3000);
    const dlg = page.locator('ytcp-dialog:visible, tp-yt-paper-dialog:visible').filter({ hasText: 'Update ad settings' }).first();
    if (!(await dlg.count())) throw new Error('no "Update ad settings?" confirmation on ' + videoId);
    await dlg.locator('ytcp-button:has-text("Update")').last().click({ timeout: 20000 });
    await page.waitForTimeout(7000);
    // and mid-roll while we are here, when the video is long enough to carry one
    const lit = page.locator('ytcp-checkbox-lit').filter({ hasText: MIDROLL_LABEL }).first();
    if (await lit.count()) {
      const box = lit.locator('#checkbox').first();
      if ((await box.getAttribute('aria-checked')) !== 'true') {
        await lit.scrollIntoViewIfNeeded();
        const bb = await box.boundingBox();
        if (bb) { await page.mouse.click(bb.x + bb.width / 2, bb.y + bb.height / 2); await page.waitForTimeout(1500); }
      }
      const save = page.locator('ytcp-button:has-text("Save"):not([disabled]), #save-button:not([disabled])').first();
      if (await save.count()) { await save.click({ timeout: 25000 }); await page.waitForTimeout(8000); }
    }
  } finally { page.off('request', grab); }
  if (!upd) throw new Error('no metadata_update captured from ' + videoId);
  return upd;
}

// The 923 pre-2024 videos on legacy ad settings: display banners only, no video ads at all. Two
// replayed writes each, then read back on hasSkippableVideoAds, which is what the conversion buys.
async function legacy(page, ids, dry) {
  const done = {}, errors = {};
  let todo = ids.slice();
  if (dry) return { dry: true, would: todo.length, sample: todo.slice(0, 5) };
  for (let round = 1; round <= 3 && todo.length; round++) {
    let upd = null, since = 0;
    for (const id of todo) {
      if (!upd || since >= REMINT_AFTER) {
        try { upd = await mintLegacy(page, id); since = 0; done[id] = 'ui-save'; continue; }
        catch (e) {
          if (e.notLegacy) { done[id] = 'already-modern'; upd = null; continue; }
          errors[id] = 'mint: ' + e.message.slice(0, 160); upd = null; continue;
        }
      }
      try {
        await replay(page, upd, id, LEGACY_FORMATS);
        await replay(page, upd, id, MIDROLL);      // refused by YouTube under 8 minutes; harmless
        done[id] = 'replay'; since++;
      } catch (e) { errors[id] = 'replay: ' + e.message.slice(0, 160); }
    }
    const back = await readBack(page, todo);
    todo = todo.filter((id) => !videoAdsOn(back[id]));
    if (!todo.length) break;
  }
  return { done, errors, stillLegacy: todo };
}

// The ad settings Studio itself reports for these ids. This is the only proof any write took:
// metadata_update answers 200 whether or not it did anything (Rs8xHbD5miQ, 20 Sep 2026).
async function readBack(page, ids) {
  const long = await listTab(page, 'upload');
  const want = new Set(ids), out = {};
  for (const r of long) if (want.has(r.id)) out[r.id] = r.adSettings || {};
  return out;
}

const midrollOn = (ad) => ((ad || {}).adBreaks || {}).hasMidrollAds === true;
const videoAdsOn = (ad) => ((ad || {}).adFormats || {}).hasSkippableVideoAds === true;

async function fix(page, ids, dry) {
  const done = {}, errors = {}, notEligible = [];
  let todo = ids.slice();
  if (dry) return { dry: true, would: todo.length, sample: todo.slice(0, 5) };
  for (let round = 1; round <= 3 && todo.length; round++) {
    let upd = null, since = 0;
    for (const id of todo) {
      if (!upd || since >= REMINT_AFTER) {
        try { upd = await mint(page, id); since = 0; done[id] = 'ui-save'; continue; }
        catch (e) {
          if (e.notEligible) { notEligible.push(id); upd = null; continue; }
          if (e.alreadyOn) { done[id] = 'already-on'; upd = null; continue; }   // still needs a mint from the next one
          errors[id] = 'mint: ' + e.message.slice(0, 160); upd = null; continue;
        }
      }
      try { const r = await replay(page, upd, id); done[id] = 'replay ' + r.status; since++; }
      catch (e) { errors[id] = 'replay: ' + e.message.slice(0, 160); }
    }
    const back = await readBack(page, todo);          // the only proof that counts
    const skip = new Set(notEligible);
    todo = todo.filter((id) => !midrollOn(back[id]) && !skip.has(id));
    if (!todo.length) break;
  }
  return { done, errors, notEligible, stillOff: todo };
}

/* ---------- main ---------- */

(async () => {
  const a = args();
  setTimeout(() => { console.log(JSON.stringify({ error: 'watchdog' })); process.exit(1); }, WATCHDOG_MS);
  const { ctx, page } = await open();
  const finish = async (obj) => {
    console.log(JSON.stringify(obj, null, 1));
    await page.close({ runBeforeUnload: false }).catch(() => {});
    await ctx.close().catch(() => {});
    process.exit(obj && obj.error ? 1 : 0);
  };
  try {
    if (a.mode === 'audit') {
      const r = await audit(page);
      if (a.out) fs.writeFileSync(a.out, JSON.stringify(r, null, 1));
      const pub = r.long.filter((v) => v.privacy === 'VIDEO_PRIVACY_PUBLIC');
      const elig = pub.filter((v) => v.seconds >= MIDROLL_MIN_SECONDS);
      const off = elig.filter((v) => !(((v.adSettings || {}).adBreaks || {}).hasMidrollAds));
      return finish({ long: r.long.length, shorts: r.shorts.length, publicLong: pub.length,
                      midrollEligible: elig.length, midrollOff: off.length, out: a.out || null });
    }
    if (a.mode === 'legacy') {
      if (!a.ids) return finish({ error: '--ids FILE required' });
      let ids = JSON.parse(fs.readFileSync(a.ids, 'utf8'));
      if (a.limit) ids = ids.slice(0, a.limit);
      return finish(await legacy(page, ids, a.dry));
    }
    if (a.mode === 'fix' || a.mode === 'verify') {
      if (!a.ids) return finish({ error: '--ids FILE required' });
      let ids = JSON.parse(fs.readFileSync(a.ids, 'utf8'));
      if (a.limit) ids = ids.slice(0, a.limit);
      if (a.mode === 'verify') {
        const back = await readBack(page, ids);
        return finish({ checked: ids.length,
                        midrollOn: ids.filter((i) => midrollOn(back[i])).length,
                        videoAdsOn: ids.filter((i) => videoAdsOn(back[i])).length,
                        stillOff: ids.filter((i) => !midrollOn(back[i]) && !videoAdsOn(back[i])) });
      }
      return finish(await fix(page, ids, a.dry));
    }
    return finish({ error: 'usage: youtube_ads.js audit|fix|legacy|verify' });
  } catch (e) { return finish({ error: e.message.slice(0, 300) }); }
})();

File: scripts/content-engine/youtube_ads.py
#!/usr/bin/env python3
"""youtube_ads.py — the ad BREAKS on a YouTube video, on top of the master switch in youtube_studio.py.

Kevin, 20 Sep 2026: "every single video has the ads switched on so that they can generate money ... that
money goes into Runpreneur's fundraising figure". Audited that day: every public video's master switch was
already On, and 817 of the 888 videos over 8 minutes had MID-ROLL ads off, so the ad in the middle of the
episode never ran. YouTube had already computed the break point on 695 of them.

Mid-roll is the biggest share of long-form ad revenue, and the YouTube Data API cannot set it, so the work
happens in Studio through the robot browser. youtube_ads.js holds the browser half and the read-back.

  youtube_ads.py midroll --video ID     # switch mid-roll (and pre/post-roll) On, then read it back
  youtube_ads.py audit --out FILE       # every video's ad breaks and ad formats
  youtube_ads.py selftest
"""
import argparse, json, os, re, subprocess, sys, tempfile

HERE = os.path.dirname(os.path.abspath(__file__))
JS = os.path.join(HERE, "youtube_ads.js")
MIDROLL_MIN_SECONDS = 480          # YouTube shows no mid-roll checkbox below 8 minutes
# What a run can say about one video. "not-eligible" is a settled answer, not a failure: a 6-minute
# episode can never carry a mid-roll, and must never sit in the backlog waiting for one.
ON, NOT_ELIGIBLE, FAILED = "on", "not-eligible", "failed"


def video_id(entry, post):
    """The 11-character YouTube id for a published post.

    The direct-upload route puts it straight on the post. The GoHighLevel route puts ITS OWN post id there
    (`6aa1688ddff4e2c4711c8b0b` on episode 2054), which is not a YouTube id and cannot be opened in Studio.
    Before 20 Sep 2026 the monetisation loop filtered on `route == "api"` and so stepped straight over every
    GHL upload without a word: 2054's episode and Short and 2195's episode were never once checked.
    The episode's own `youtube_link` carries the real id for the long video, so use it."""
    # An id written on the post by hand, for the one case nothing can derive: a Short uploaded through
    # GoHighLevel, whose only record of the YouTube video is GHL's own post id (episode 2054, 20 Sep 2026,
    # matched on the channel as WUTZvEoLDbQ). `id` is left alone because the GHL status sync matches on it.
    fixed = (post or {}).get("youtube_video_id") or ""
    if re.fullmatch(r"[A-Za-z0-9_-]{11}", fixed):
        return fixed
    vid = (post or {}).get("id") or ""
    if re.fullmatch(r"[A-Za-z0-9_-]{11}", vid):
        return vid
    if (post or {}).get("clip") == "full":
        m = re.search(r"(?:youtu\.be/|v=)([A-Za-z0-9_-]{11})", (entry or {}).get("youtube_link") or "")
        if m: return m.group(1)
    return None


def _node():
    for c in (os.environ.get("NODE"), subprocess.run(["bash", "-lc", "command -v node"], capture_output=True, text=True).stdout.strip()):
        if c and os.path.exists(c): return c
    home = os.path.expanduser("~/.nvm/versions/node")
    vers = sorted([d for d in os.listdir(home) if d.startswith("v")], key=lambda v: [int(x) for x in re.findall(r"\d+", v)]) if os.path.isdir(home) else []
    if vers: return os.path.join(home, vers[-1], "bin", "node")
    return "node"


def _run(argv, timeout=7500):      # longer than youtube_ads.js's own two-hour watchdog, so the JS reports first
    r = subprocess.run([_node(), JS] + argv, capture_output=True, text=True, timeout=timeout)
    out = (r.stdout or "").strip()
    if "{" not in out: raise SystemExit("youtube_ads.js gave nothing back: %s" % ((r.stderr or out)[-240:]))
    return json.loads(out[out.index("{"):])


def midroll(ids):
    """Switch mid-roll On for these video ids. Returns {id: "on"|"not-eligible"|"failed"}.

    A 200 from Studio's write is NOT proof — without the page's BotGuard attestation the endpoint accepts
    the call and changes nothing (proved on Rs8xHbD5miQ, 20 Sep 2026). youtube_ads.js reads every id back
    off Studio's own list before it reports, so anything not proved On comes back as failed."""
    ids = [i for i in ids if re.fullmatch(r"[A-Za-z0-9_-]{11}", i or "")]
    if not ids: return {}
    with tempfile.NamedTemporaryFile("w", suffix=".json", delete=False) as fh:
        json.dump(ids, fh); path = fh.name
    try:
        res = _run(["fix", "--ids", path])
    finally:
        os.remove(path)
    if res.get("error"): raise SystemExit("mid-roll run failed: %s" % res["error"])
    out = {}
    for i in ids:
        if i in (res.get("notEligible") or []): out[i] = NOT_ELIGIBLE
        elif i in (res.get("stillOff") or []) or i in (res.get("errors") or {}): out[i] = FAILED
        else: out[i] = ON
    return out


def audit(out_file=None):
    return _run(["audit"] + (["--out", out_file] if out_file else []))


def selftest():
    api = {"id": "yGOws0WofyU", "clip": "full", "route": "api"}
    ghl_full = {"id": "6aa1688ddff4e2c4711c8b0b", "clip": "full", "route": None}
    ghl_short = {"id": "6aa181234bdc74b45aeab6d5", "clip": "lfmd", "route": None}
    ent = {"youtube_link": "https://youtu.be/AT0l-Ri5ZJ0"}
    assert video_id(ent, api) == "yGOws0WofyU", "a direct upload carries its own id"
    # the real episode 2054, which the old route == "api" filter skipped in silence
    assert video_id(ent, ghl_full) == "AT0l-Ri5ZJ0", "a GoHighLevel upload is resolved through the episode link"
    assert video_id(ent, ghl_short) is None, "a GHL Short has no link to resolve, so it is reported, never assumed"
    assert video_id(ent, dict(ghl_short, youtube_video_id="WUTZvEoLDbQ")) == "WUTZvEoLDbQ", "an id written on the post by hand wins"
    assert video_id(ent, dict(api, youtube_video_id="nonsense")) == "yGOws0WofyU", "a bad hand-written id is ignored, not trusted"
    assert video_id({}, ghl_full) is None and video_id({"youtube_link": "not a link"}, ghl_full) is None
    assert video_id(ent, {"id": "", "clip": "full"}) == "AT0l-Ri5ZJ0"
    assert video_id({"youtube_link": "https://www.youtube.com/watch?v=bSzyDyON4uc"}, ghl_full) == "bSzyDyON4uc"
    assert midroll([]) == {} and midroll(["not-an-id"]) == {}, "only an 11-character id ever reaches the browser"
    src = open(JS).read()
    assert "readBack" in src and "the only proof that counts" in src, "every write is read back off Studio"
    assert "attestationResponseData" in src or "attestation" in src, "the write reuses the page's minted attestation"
    assert "MIDROLL_MIN_SECONDS = 480" in src and MIDROLL_MIN_SECONDS == 480, "mid-roll needs 8 minutes"
    assert os.path.basename(_node()) == "node", "a node binary is found even under launchd's bare PATH"
    print(json.dumps({"checks": 13, "failed": []}))


if __name__ == "__main__":
    ap = argparse.ArgumentParser(); ap.add_argument("mode"); ap.add_argument("--video", default=""); ap.add_argument("--out", default="")
    a = ap.parse_args()
    if a.mode == "selftest": selftest()
    elif a.mode == "midroll": print(json.dumps(midroll([a.video])))
    elif a.mode == "audit": print(json.dumps(audit(a.out or None)))
    else: raise SystemExit("usage: youtube_ads.py midroll --video ID | audit [--out FILE] | selftest")
